README.md
# PWM Tool

![Build Status](https://github.com/tano-systems/pwm-tool/actions/workflows/build.yml/badge.svg?branch=master) ![Tests Status](https://github.com/tano-systems/pwm-tool/actions/workflows/build_and_test.yml/badge.svg?branch=master)

This is a simple Linux command line tool designed to control the available PWM channels via the sysfs interface. The main purpose of this utility is to control a buzzer device connected to the corresponding PWM channel.

## Build and Install

Clone tool sources (for example, to `~/pwm-tool` directory):

```shell
$ git clone https://github.com/tano-systems/pwm-tool.git ~/pwm-tool
```

Jump to the directory with sources:

```shell
$ cd ~/pwm-tool
```

Create a `build` subdirectory and jump to it:

```shell
$ mkdir build
$ cd build
```

Run CMake:

```shell
$ cmake ../
```

Run Make:

```shell
$ make
```

This will produce binary `pwm` in the `build` directory. To install this binary to `/usr/bin` run command as `root` user:

```shell
# make install
```

## Tests

To run unit tests use following command:

```shell
$ make build_and_test
```

## Usage

Usage syntax:

```shell
pwm [options]
```

*   `[options]` is a one or more additional optional options that are described in the "[Options](#options)" section.

### Options

| Short Form         | Long Form                  | Default Value | Description                                                  |
| ------------------ | -------------------------- | ------------- | ------------------------------------------------------------ |
| `-h`               | `--help`                   | -             | Display help and usage text.                                 |
| `-p <chip>`        | `--chip=<chip>`            | `0`           | Set PWM chip number to `<chip>`                              |
| `-c <channel>`     | `--channel=<channel>`      | `0`           | Set PWM channel number to `<channel>`                        |
| `-f <freq_hz>`     | `--frequency=<freq_hz>`    | `1000`        | Set PWM frequency in Hz. If the specified frequency is `0`, the PWM will not be enabled. |
| `-d <duration_ms>` | `--duration=<duration_ms>` | `250`         | Set PWM enabled state duration in milliseconds.              |
| `-k`               | `--keep-enabled`           | -             | If specified, PWM will remain enabled on exit.               |
| `-s <script>`      | `--script=<script>`        | -             | Run PWM commands script. See details in "[Scripts Syntax](#scripts-syntax)" section. |
| -                  | `--version`                | -             | Display PWM tool version.                                    |

### Scripts Syntax

The script consists of commands separated by one or more spaces:
```
<command1> <command2> ... <commandN>
```

Each command consists of one or more operations, which are stored one after the other without spaces:
```
<op1><op2>...<opN>
```

The operation is represented by a single character of the alphabet. Operation can have only one optional numeric argument, which must be written immediately after the operation character:
```
O[argument]
```

For example, the operation `F` with the argument 1250 should be written as follows:
```
F1250
```

And a command consisting of two operations, `F` with argument 1250 and `d` with argument 50, as follows:
```
F1250d50
```

Examples of how to write multiple commands can be seen in the "[Examples](#examples)" section.

Available PWM command operations:

| Operation | Description                                                  |
| --------- | ------------------------------------------------------------ |
| `f[hz]`   | Set frequency from optional argument. If no argument is specified, the current default frequency will be used. |
| `F[hz]`   | Same as `f[hz]`, but if an argument is specified, the value will then be used as the default frequency for subsequent commands. |
| `d[ms]`   | Set duration from optional argument. If no argument is specified, the current default duration will be used. |
| `D[ms]`   | Same as `d[ms]`, but if an argument is specified, the value will then be used as the default duration for subsequent commands. |
| `k`       | Keep the PWM enabled when the command is completed.          |

If no `f` or `F` operation is specified in a command, then a frequency of 0 will be used for that command, i.e. such commands can be used to delay script execution.

If no `d` or `D` operation is specified in a command, the current default duration will be used for that command. Changing the default duration can either be done through the configuration structure or directly at runtime with the `D[ms]` operation.


## Examples

Three short beeps with a frequency of 1000 Hz (duration 100 ms with 50 ms delay between):

```shell
$ pwm -f 1000 -d 100
$ pwm -f 0 -d 50
$ pwm -f 1000 -d 100
$ pwm -f 0 -d 50
$ pwm -f 1000 -d 100
```

Same for PWM commands script mode:
```shell
$ pwm -s "F1000D100 d50 f d50 f"
```

## Changelog

See [CHANGELOG.md](CHANGELOG.md).

## License

This work is free. You can redistribute it and/or modify it under the
terms of the Do What The Fuck You Want To Public License, Version 2,
as published by Sam Hocevar. See <http://www.wtfpl.net/> for more details.

## Maintainers

*   Anton Kikin <mailto:a.kikin@tano-systems.com>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).

All dates in this document are in `DD.MM.YYYY` format.

## [Unreleased]

**Nothing yet**

## [Version 1.0.1] (29.01.2021)

### Added
- Add support for exporting PWM channel
- Add SPDX tags to sources

## [Version 1.0.0] (26.01.2021)

Initial release

[Unreleased]: https://github.com/tano-systems/pwm-tool/tree/master
[Version 1.0.1]: https://github.com/tano-systems/pwm-tool/releases/tag/v1.0.1
[Version 1.0.0]: https://github.com/tano-systems/pwm-tool/releases/tag/v1.0.0


tests/generic/def.sh
#!/bin/sh
#
# SPDX-License-Identifier: WTFPL
# SPDX-FileCopyrightText: 2021 Anton Kikin <a.kikin@tano-systems.com>
#

function do_test {
	local SYSFS
	local ENABLE
	local PERIOD
	local DUTY_CYCLE
	local RET
	local D1
	local D2

	# Create sysfs root + chip folder + channel folder
	test_sysfs_create ${DEFAULT_PWM_CHIP} ${DEFAULT_PWM_CHANNEL} SYSFS

	D1=$(date "+%s %N")
	${PWM_TEST_BIN}
	RET=$?
	D2=$(date "+%s %N")

	local DMS=$(date_diff_ms ${D2} ${D1})

	test_assert_range \
		${DMS} \
		${DEFAULT_PWM_DURATION_MS} \
		$((${DEFAULT_PWM_DURATION_MS} + 100)) \
		"execution duration"

	test_assert_eq "${RET}" "${PWM_E_OK}" "return code"

	test_sysfs_read ${SYSFS} ENABLE PERIOD DUTY_CYCLE

	test_assert_eq "${ENABLE}" "10" "enable data check"
	test_assert_eq "${PERIOD}" "1000000" "period data check"
	test_assert_eq "${DUTY_CYCLE}" "500000" "duty_cycle data check"

	test_passed
}

. ${PWM_TEST_ROOT}/pwm-test-common.sh.inc


tests/generic/version.sh
#!/bin/sh
#
# SPDX-License-Identifier: WTFPL
# SPDX-FileCopyrightText: 2021 Anton Kikin <a.kikin@tano-systems.com>
#

function do_test {
	local RET
	local SYSFS
	local ENABLE
	local PERIOD
	local DUTY_CYCLE

	# Create sysfs root + chip folder + channel folder
	test_sysfs_create ${DEFAULT_PWM_CHIP} ${DEFAULT_PWM_CHANNEL} SYSFS

	STDOUT=$(${PWM_TEST_BIN} --version)
	RET=$?

	test_assert_eq "${RET}" "${PWM_E_OK}" "return code"
	test_assert_eq "${STDOUT}" "${PWM_VERSION}" "stdout contents"

	test_sysfs_read ${SYSFS} ENABLE PERIOD DUTY_CYCLE

	test_assert_eq "${ENABLE}" "" "enable data check"
	test_assert_eq "${PERIOD}" "" "period data check"
	test_assert_eq "${DUTY_CYCLE}" "" "duty_cycle data check"

	test_passed
}

. ${PWM_TEST_ROOT}/pwm-test-common.sh.inc


tests/script/script-01.sh
#!/bin/sh
#
# SPDX-License-Identifier: WTFPL
# SPDX-FileCopyrightText: 2021 Anton Kikin <a.kikin@tano-systems.com>
#

function do_test {
	local RET
	local SYSFS
	local ENABLE
	local PERIOD
	local DUTY_CYCLE
	local D1
	local D2

	# Create sysfs root + chip folder + channel folder
	test_sysfs_create ${DEFAULT_PWM_CHIP} ${DEFAULT_PWM_CHANNEL} SYSFS

	D1=$(date "+%s %N")
	${PWM_TEST_BIN} --script="F1000D100 d50 f d50 f"
	RET=$?
	D2=$(date "+%s %N")

	test_assert_eq "${RET}" "${PWM_E_OK}" "return code"

	local DMS=$(date_diff_ms ${D2} ${D1})

	# expected duration is 400 ms (3 x 100 + 2 x 50)
	test_assert_range ${DMS} 400 500 "execution duration"

	test_sysfs_read ${SYSFS} ENABLE PERIOD DUTY_CYCLE

	local EXP_ENABLE="10010010"
	local EXP_PERIOD="100000010000001000000"
	local EXP_DUTY_CYCLE="50000005000000500000"

	test_assert_eq "${ENABLE}" "${EXP_ENABLE}" "enable data check"
	test_assert_eq "${PERIOD}" "${EXP_PERIOD}" "period data check"
	test_assert_eq "${DUTY_CYCLE}" "${EXP_DUTY_CYCLE}" "duty_cycle data check"

	test_passed
}

. ${PWM_TEST_ROOT}/pwm-test-common.sh.inc


src/pwm.c
/*
 * SPDX-License-Identifier: WTFPL
 * SPDX-FileCopyrightText: 2021 Anton Kikin <a.kikin@tano-systems.com>
 *
 * PWM tool
 * Copyright © 2021 Anton Kikin <a.kikin@tano-systems.com>
 *
 * This work is free. You can redistribute it and/or modify it under the
 * terms of the Do What The Fuck You Want To Public License, Version 2,
 * as published by Sam Hocevar. See the COPYING file for more details.
 */

/**
 * @file
 * @brief PWM tool source file
 *
 * @author Anton Kikin <a.kikin@tano-systems.com>
 */

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <ctype.h>        /* isspace(), isdigit() */
#include <errno.h>        /* EINTR */
#include <time.h>         /* clock_nanosleep() */
#include <fcntl.h>        /* openat() */
#include <math.h>         /* roundf() */
#include <linux/limits.h> /* NAME_MAX */

#include "pwm.h"

/* ----------------------------------------------------------------------- */

#ifndef SYSFS_PWM_ROOT

/**
 * Root in sysfs for PWM control
 *
 * Full sysfs path for specified PWM channel is:
 * <code>
 * sprintf(path, SYSFS_PWM_ROOT"/"
 *               SYSFS_PWM_CHIP_FOLDER_FMT"/"
 *               SYSFS_PWM_CH_FOLDER_FMT,
 *               chip, ch);
 * </code>
 */
#define SYSFS_PWM_ROOT  "/sys/class/pwm"
#endif

#ifndef SYSFS_PWM_CHIP_FOLDER_FMT

/**
 * Format for PWM chip number subfolder in sysfs
 *
 * @see SYSFS_PWM_ROOT
 */
#define SYSFS_PWM_CHIP_FOLDER_FMT  "pwmchip%u"
#endif

#ifndef SYSFS_PWM_CH_FOLDER_FMT

/**
 * Format for PWM channel number subfolder in sysfs
 *
 * @see SYSFS_PWM_ROOT
 */
#define SYSFS_PWM_CH_FOLDER_FMT  "pwm%u"
#endif

#ifndef SYSFS_PWM_FILE_ENABLE

/** File name in sysfs for control enabled state of the PWM */
#define SYSFS_PWM_FILE_ENABLE  "enable"
#endif

#ifndef SYSFS_PWM_FILE_PERIOD

/** File name in sysfs for control period of the PWM */
#define SYSFS_PWM_FILE_PERIOD  "period"
#endif

#ifndef SYSFS_PWM_FILE_DUTY_CYCLE

/** File name in sysfs for control duty-cycle of the PWM */
#define SYSFS_PWM_FILE_DUTY_CYCLE  "duty_cycle"
#endif

/* ----------------------------------------------------------------------- */

static pwm_status_t pwm_export(
	int chip_fd,
	unsigned int channel
)
{
	int fd_export = openat(chip_fd, "export", O_WRONLY);
	if (fd_export < 0)
		return PWM_E_EXPORT_FAILED;

	char chnum[16];
	ssize_t size;
	size = snprintf(chnum, sizeof(chnum), "%u", channel);

	if (write(fd_export, chnum, size) != size) {
		close(fd_export);
		return PWM_E_EXPORT_FAILED;
	}

	close(fd_export);
	return PWM_E_OK;
}

static pwm_status_t pwm_state_read(pwm_t *pwm)
{
	char buffer[16];
	ssize_t size;

	size = read(pwm->fd_enable, buffer, sizeof(buffer));
	if (size <= 0)
		return PWM_E_IO;

	buffer[size] = '\0';
	pwm->enabled = !!strtoul(buffer, NULL, 10);

	size = read(pwm->fd_period, buffer, sizeof(buffer));
	if (size <= 0)
		return PWM_E_IO;

	buffer[size] = '\0';
	pwm->period = (unsigned int)strtoul(buffer, NULL, 10);

	size = read(pwm->fd_dutycycle, buffer, sizeof(buffer));
	if (size <= 0)
		return PWM_E_IO;

	buffer[size] = '\0';
	pwm->duty_cycle = (unsigned int)strtoul(buffer, NULL, 10);

	return PWM_E_OK;
}

/* ----------------------------------------------------------------------- */

pwm_status_t pwm_open(
	pwm_t *pwm,
	unsigned int chip,
	unsigned int channel,
	unsigned int flags
)
{
	int pwm_root_fd;
	int pwm_chip_fd;
	int pwm_channel_fd;
	pwm_status_t ret;

	char filename[NAME_MAX];

	memset(pwm, 0, sizeof(pwm_t));

	pwm->chip = chip;
	pwm->channel = channel;
	pwm->flags = flags;

	/* Open sysfs root */
	pwm_root_fd = open(SYSFS_PWM_ROOT,
		O_PATH | O_DIRECTORY);

	if (pwm_root_fd < 0)
		return PWM_E_NO_SYSFS;

	/* Open PWM chip folder */
	snprintf(filename, sizeof(filename),
		SYSFS_PWM_CHIP_FOLDER_FMT, chip);
	pwm_chip_fd = openat(pwm_root_fd, filename,
		O_PATH | O_DIRECTORY);

	if (pwm_chip_fd < 0) {
		close(pwm_root_fd);
		return PWM_E_NO_CHIP;
	}

	close(pwm_root_fd);

	/* Open PWM channel folder */
	snprintf(filename, sizeof(filename),
		SYSFS_PWM_CH_FOLDER_FMT, channel);
	pwm_channel_fd = openat(pwm_chip_fd, filename,
		O_PATH | O_DIRECTORY);

	if (pwm_channel_fd < 0) {
		if (pwm->flags & PWM_FLAG_EXPORT) {
			ret = pwm_export(
				pwm_chip_fd, pwm->channel);

			if (ret == PWM_E_OK) {
				/* Try to open channel again after exporting */
				pwm_channel_fd = openat(pwm_chip_fd, filename,
					O_PATH | O_DIRECTORY);
			}
		}

		if (pwm_channel_fd < 0) {
			close(pwm_chip_fd);
			return PWM_E_NO_CHANNEL;
		}
	}

	close(pwm_chip_fd);

	/* Open control files */
	pwm->fd_enable = openat(pwm_channel_fd,
		SYSFS_PWM_FILE_ENABLE, O_RDWR);

	if (pwm->fd_enable < 0)
		goto failed;

	pwm->fd_dutycycle = openat(pwm_channel_fd,
		SYSFS_PWM_FILE_DUTY_CYCLE, O_RDWR);

	if (pwm->fd_dutycycle < 0)
		goto failed;

	pwm->fd_period = openat(pwm_channel_fd,
		SYSFS_PWM_FILE_PERIOD, O_RDWR);

	if (pwm->fd_period < 0)
		goto failed;

	ret = pwm_state_read(pwm);
	if (ret != PWM_E_OK)
		goto failed;

	close(pwm_channel_fd);
	return ret;

failed:
	if (pwm_channel_fd > 0)
		close(pwm_channel_fd);

	if (pwm->fd_enable > 0)
		close(pwm->fd_enable);

	if (pwm->fd_dutycycle > 0)
		close(pwm->fd_dutycycle);

	if (pwm->fd_period > 0)
		close(pwm->fd_period);

	return PWM_E_IO;
}

static pwm_status_t pwm_enable_ext(
	pwm_t *pwm,
	unsigned int period,
	unsigned int duty)
{
	ssize_t len;
	char buf[32];

	/*
	 * Temporarily set a minimum duty-cycle to be able
	 * to set any period, even one that is larger than
	 * the current set duty-cycle.
	 */
	if (pwm->period > 0) {
		if (write(pwm->fd_dutycycle, "0", 1) != 1)
			return PWM_E_IO;
	}

	len = snprintf(buf, sizeof(buf), "%u", period);
	if (write(pwm->fd_period, buf, len) != len)
		return PWM_E_IO;

	pwm->period = period;

	/* Set specified duty-cycle */
	len = snprintf(buf, sizeof(buf), "%u", duty);
	if (write(pwm->fd_dutycycle, buf, len) != len)
		return PWM_E_IO;

	pwm->period = duty;

	if (write(pwm->fd_enable, "1", 1) != 1)
		return PWM_E_IO;

	pwm->enabled = 1;

	return PWM_E_OK;
}

pwm_status_t pwm_enable(pwm_t *pwm, unsigned int freq)
{
	float fresult;
	unsigned int period;
	unsigned int duty;

	if ((freq < 1) || (freq > 500000000))
		return PWM_E_INVALID_FREQ;

	fresult = 1000000000 / (float)freq;
	period  = (unsigned int)(roundf(fresult));
	duty    = (unsigned int)(roundf(period / 2.0f));

	return pwm_enable_ext(pwm, period, duty);
}

pwm_status_t pwm_disable(pwm_t *pwm)
{
	if (write(pwm->fd_enable, "0", 1) != 1)
		return PWM_E_IO;

	pwm->enabled = 0;

	return PWM_E_OK;
}

pwm_status_t pwm_close(pwm_t *pwm)
{
	if (pwm->fd_enable > 0)
		close(pwm->fd_enable);

	if (pwm->fd_dutycycle > 0)
		close(pwm->fd_dutycycle);

	if (pwm->fd_period > 0)
		close(pwm->fd_period);

	return PWM_E_OK;
}

static pwm_status_t pwm_delay_abs_time(
	pwm_t *pwm,
	const struct timespec *ts,
	struct timespec *remain
)
{
	int ret;

	ret = clock_nanosleep(CLOCK_MONOTONIC, TIMER_ABSTIME, ts, remain);
	if (ret == 0)
		return PWM_E_OK;

	if (ret == EINTR)
		return PWM_E_INTR;
	else
		return PWM_E_FAILED;
}

const char *pwm_strstatus(const pwm_status_t status)
{
	switch (status) {
		case PWM_E_OK:
			return "Ok";

		case PWM_E_NO_SYSFS:
			return "PWM sysfs interface is not available";

		case PWM_E_NO_CHIP:
			return "PWM chip is not available";

		case PWM_E_NO_CHANNEL:
			return "PWM channel is not available";

		case PWM_E_INVALID_FREQ:
			return "Invalid frequency";

		case PWM_E_INVALID_DURATION:
			return "Invalid duration";

		case PWM_E_INTR:
			return "Interrupted";

		case PWM_E_FAILED:
			return "General failure";

		case PWM_E_EXPORT_FAILED:
			return "Exporting failure";

		default:
			return "Unknown";
	}
}

/* ----------------------------------------------------------------------- */

/**
 * PWM command data structure
 */
typedef struct {
	/** Frequency */
	unsigned int frequency_hz;

	/** Duration */
	unsigned int duration_ms;

	/** Keep enabled after command executed */
	int keep_enabled;

	/** Timestamp when command must be finished */
	struct timespec ts_end;

} pwm_cmd_t;

/**
 * PWM commands fetcher data structure
 */
typedef struct {
	const char *script;
	const char *pos;

	/** Default frequency (used if not specified in command) */
	unsigned int frequency_hz;

	/** Default duration (used if not specified in command) */
	unsigned int duration_ms;

} pwm_cmd_fetcher_t;

/**
 * Initialize PWM commands fetcher
 */
static void pwm_cmd_fetch_init(
	pwm_cmd_fetcher_t *f,
	const char *script,
	unsigned int frequency_hz,
	unsigned int duration_ms
)
{
	f->script = f->pos = script;

	f->frequency_hz = frequency_hz;
	f->duration_ms  = duration_ms;
}

/**
 * Fetch single PWM command
 *
 * @return -1 Command successufully fetched.
 * @return  0 All commands are fetched, no more commands available.
 * @return <0 Syntax error or invalid command.
 */
static int pwm_cmd_fetch(pwm_cmd_fetcher_t *f, pwm_cmd_t *cmd)
{
	/* Skip whitespaces */
	while (*(f->pos) && isspace(*(f->pos)))
		f->pos++;

	if (!*(f->pos))
		return 0;

	cmd->keep_enabled = 0;
	cmd->frequency_hz = 0;
	cmd->duration_ms  = f->duration_ms;

	/* Parse operations */
	while (f->pos[0] && !isspace(f->pos[0])) {
		char op = f->pos[0];
		switch(op) {
			case 'k':
				cmd->keep_enabled = 1;
				f->pos++;
				break;

			case 'f': /* fallthrough */
			case 'F':
				if (isdigit(f->pos[1])) {
					cmd->frequency_hz =
						(unsigned int)strtoul(f->pos + 1, (char **)&f->pos, 10);

					if (op == 'F')
						f->frequency_hz = cmd->frequency_hz;
				}
				else {
					cmd->frequency_hz = f->frequency_hz;
					f->pos++;
				}
				break;

			case 'd': /* fallthrough */
			case 'D':
				if (isdigit(f->pos[1])) {
					cmd->duration_ms =
						(unsigned int)strtoul(f->pos + 1, (char **)&f->pos, 10);

					if (op == 'D')
						f->duration_ms = cmd->duration_ms;
				}
				else {
					cmd->duration_ms = f->duration_ms;
					f->pos++;
				}
				break;

			default:
				fprintf(stderr,
					"ERROR: Unknown command '%c' in script at position %u\n",
					f->pos[0],
					(unsigned int)(f->pos - f->script) + 1);

				return -1;
		}
	}

	return 1;
}

/**
 * Execute single PWM command
 */
static int pwm_cmd_execute(pwm_t *pwm, pwm_cmd_t *cmd)
{
	pwm_status_t ret = PWM_E_OK;

	if (cmd->frequency_hz) {
		ret = pwm_enable(pwm, cmd->frequency_hz);
		if (ret != PWM_E_OK) {
			fprintf(stderr,
				"ERROR: Can't enable PWM channel %u of chip %u: %s\n",
				pwm->channel, pwm->chip, pwm_strstatus(ret));
			return ret;
		}
	}
	else {
		ret = pwm_disable(pwm);
	}

	/* Sleep for specified duration */
	if (cmd->duration_ms)
		ret = pwm_delay_abs_time(pwm, &cmd->ts_end, NULL);

	if (!cmd->keep_enabled && cmd->frequency_hz)
		ret = pwm_disable(pwm);

	return ret;
}

pwm_status_t pwm_execute(
	pwm_t *pwm,
	const pwm_execute_config_t *config)
{
	pwm_status_t ret = PWM_E_OK;
	pwm_cmd_t cmd;
	pwm_cmd_fetcher_t fetcher;
	int fetched;

	pwm_cmd_fetch_init(
		&fetcher,
		config->script,
		config->default_frequency_hz,
		config->default_duration_ms
	);

	clock_gettime(CLOCK_MONOTONIC, &cmd.ts_end);

	while (1) {
		if (config->stop_flag && *(config->stop_flag))
			break;

		fetched = pwm_cmd_fetch(&fetcher, &cmd);
		if (fetched < 0) {
			pwm_disable(pwm);
			return PWM_E_FAILED;
		}
		
		if (!fetched)
			return PWM_E_OK;

		cmd.ts_end.tv_sec  += cmd.duration_ms / 1000;
		cmd.ts_end.tv_nsec += (cmd.duration_ms % 1000) * 1000000L;

		if (cmd.ts_end.tv_nsec >= 1000000000L) {
			cmd.ts_end.tv_nsec -= 1000000000L;
			cmd.ts_end.tv_sec++;
		}

		ret = pwm_cmd_execute(pwm, &cmd);
		if (ret != PWM_E_OK)
			break;
	}

	return ret;
}


src/main.c
/*
 * SPDX-License-Identifier: WTFPL
 * SPDX-FileCopyrightText: 2021 Anton Kikin <a.kikin@tano-systems.com>
 *
 * PWM tool
 * Copyright © 2021 Anton Kikin <a.kikin@tano-systems.com>
 *
 * This work is free. You can redistribute it and/or modify it under the
 * terms of the Do What The Fuck You Want To Public License, Version 2,
 * as published by Sam Hocevar. See the COPYING file for more details.
 */

/**
 * @file
 * @brief PWM tool main source file
 *
 * @author Anton Kikin <a.kikin@tano-systems.com>
 */

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <errno.h>

#include "pwm.h"

/* ----------------------------------------------------------------------- */

#ifndef DEFAULT_PWM_CHIP

/** Default PWM chip number */
#define DEFAULT_PWM_CHIP  0
#endif

#ifndef DEFAULT_PWM_CHANNEL

/** Default PWM channel number */
#define DEFAULT_PWM_CHANNEL  0
#endif

#ifndef DEFAULT_PWM_FREQUENCY_HZ

/** Default frequency in Hz */
#define DEFAULT_PWM_FREQUENCY_HZ  1000
#endif

#ifndef DEFAULT_PWM_DURATION_MS

/** Default duration in milliseconds */
#define DEFAULT_PWM_DURATION_MS  250
#endif

/* ----------------------------------------------------------------------- */

/**
 * @brief Configuration data structure
 */
typedef struct config {

	/** PWM chip number.
	 *  Default value specified in @ref DEFAULT_PWM_CHIP. */
	unsigned int chip;

	/** PWM channel number.
	 *  Default value specified in @ref DEFAULT_PWM_CHANNEL. */
	unsigned int channel;

	/** PWM frequency in Hz
	 *  Default value specified in @ref DEFAULT_PWM_FREQUENCY_HZ. */
	unsigned int frequency_hz;

	/** PWM duration in ms
	 *  Default value specified in @ref DEFAULT_PWM_DURATION_MS. */
	unsigned int duration_ms;

	/** If set, PWM will remain enabled on exit. */
	int keep_enabled;

	char *script;

} config_t;

/* ----------------------------------------------------------------------- */

/** Global exit flag (used in script mode) */
static int exit_flag = 0;

/**
 * @brief Global configuration structure
 */
static config_t config = {
	.chip         = DEFAULT_PWM_CHIP,
	.channel      = DEFAULT_PWM_CHANNEL,
	.frequency_hz = DEFAULT_PWM_FREQUENCY_HZ,
	.duration_ms  = DEFAULT_PWM_DURATION_MS,
	.keep_enabled = 0,
};

/**
 * @brief Short command line options list
 */
static const char *opts_str = "hp:c:f:d:s:k";

/**
 * @brief Long command line options list
 */
static const struct option opts[] = {
	{ .name = "help",         .val = 'h' },
	{ .name = "chip",         .val = 'p', .has_arg = 1 },
	{ .name = "channel",      .val = 'c', .has_arg = 1 },
	{ .name = "frequency",    .val = 'f', .has_arg = 1 },
	{ .name = "duration",     .val = 'd', .has_arg = 1 },
	{ .name = "script",       .val = 's', .has_arg = 1 },
	{ .name = "keep-enabled", .val = 'k' },
	{ .name = "version",      .val = 'V' },
	{ 0 }
};

/**
 * Display program usage help
 */
static void display_usage(void)
{
	fprintf(stdout,
		"\n"
		"PWM tool " PWM_VERSION "\n"
		"Copyright (c) 2021 Anton Kikin <a.kikin@tano-systems.com>\n"
		"\n"
		"Usage: pwm [options]\n"
		"\n"
		"Options:\n"
		"  -h, --help\n"
		"        Display this help text.\n"
		"\n"
		"  -p, --chip <chip>\n"
		"        Select PWM chip number.\n"
		"        Default: %u\n"
		"\n"
		"  -c, --channel <channel>\n"
		"        Select PWM chip channel number.\n"
		"        Default: %u\n"
		"\n"
		"  -f, --frequency <frequency_in_hz>\n"
		"        Set PWM frequency in Hz.\n"
		"        Default: %u\n"
		"\n"
		"  -d, --duration <duration_in_ms>\n"
		"        Set PWM duration in milliseconds.\n"
		"        Default: %u\n"
		"\n"
		"  -k, --keep-enabled\n"
		"        If specified, PWM will remain enabled on exit.\n"
		"        Default: disable PWM on exit\n"
		"\n"
		"  -s, --script <script>\n"
		"        Run PWM commands script.\n"
		"\n"
		"  --version\n"
		"        Display PWM tool version.\n"
		"\n",
		DEFAULT_PWM_CHIP,
		DEFAULT_PWM_CHANNEL,
		DEFAULT_PWM_FREQUENCY_HZ,
		DEFAULT_PWM_DURATION_MS
	);
}

/**
 * Parse command line arguments into @ref config global structure
 *
 * @param[in] argc  Number of arguments
 * @param[in] argv  Array of the pointers to the arguments
 *
 * @return 0 on success
 * @return <0 on error
 */
static int parse_cli_args(int argc, char *argv[])
{
	int opt;

	while((opt = getopt_long(argc, argv, opts_str, opts, NULL)) != EOF) {
		switch(opt) {
			case '?': /* Invalid option */
				return -EINVAL;

			case 'h': /* --help */
				display_usage();
				exit(0);

			case 'p': /* --chip */
				config.chip =
					(unsigned int)strtoul(optarg, NULL, 0);
				break;

			case 'c': /* --channel */
				config.channel =
					(unsigned int)strtoul(optarg, NULL, 0);
				break;

			case 'f': /* --frequency */
				config.frequency_hz =
					(unsigned int)strtoul(optarg, NULL, 0);
				break;

			case 'd': /* --duration */
				config.duration_ms =
					(unsigned int)strtoul(optarg, NULL, 0);
				break;

			case 's': /* --script */
				config.script = strdup(optarg);
				if (!config.script) {
					fprintf(stderr, "ERROR: Out of memory");
					exit(-ENOMEM);
				}
				break;

			case 'k': /* --keep-enabled */
				config.keep_enabled = 1;
				break;

			case 'V': /* --version */
				fprintf(stdout, "%s\n", PWM_VERSION);
				exit(0);

			default:
				break;
		}
	}

	return 0;
}

/**
 * Signal handler
 */
static void handle_signal(int signal)
{
	exit_flag = 1;
}

/**
 * Cleanup
 */
void cleanup(void)
{
	if (config.script)
		free(config.script);
}

/**
 * Program start point
 *
 * @param[in] argc  Number of arguments
 * @param[in] argv  Array of the pointers to the arguments
 *
 * @return 0 on success
 * @return <0 on error
 */
int main(int argc, char *argv[])
{
	pwm_status_t ret = 0;
	pwm_t pwm;

	if (parse_cli_args(argc, argv)) {
		display_usage();
		return EINVAL;
	}

	atexit(cleanup);

	signal(SIGINT, handle_signal);

	ret = pwm_open(&pwm, config.chip, config.channel, PWM_FLAG_EXPORT);
	if (ret != PWM_E_OK) {
		fprintf(stderr,
			"ERROR: Can't open PWM channel %u of chip %u: %s\n",
			config.channel, config.chip, pwm_strstatus(ret));
		exit(ret);
	}

	pwm_execute_config_t pwm_execute_config = {
		.script               =  config.script,
		.default_frequency_hz =  config.frequency_hz,
		.default_duration_ms  =  config.duration_ms,
		.stop_flag            = &exit_flag,
	};

	if (!config.script) {
		if (config.keep_enabled)
			pwm_execute_config.script = "fdk";
		else
			pwm_execute_config.script = "fd";
	}

	ret = pwm_execute(&pwm, &pwm_execute_config);

	pwm_close(&pwm);

	exit(ret);
}
